### DISCUSSION TITLE
Storage: server-side upload and download

### DISCUSSION THREAD

**Discussion by joshnuss** [supabase/supabase#1002] [Feature Requests] [answered]
Really liking Storage!

It would be handy for `supabase.storage.upload()` and `supabase.storage.download()` to work server side too.  Currently there's a check if `window` is defined.

This would be useful for syncing a folder to storage, or migrating from another provider.

We can accomplish this by allowing the function to accept a `File` in-browser and a `String` (path) server-side.
And maybe while we're at it, could be worth adding a `downloadBlob` and `uploadBlob` that works without the filesystem.

Let me know if you want a PR, happy to help.

**Comment by libovness** [comment]
+1 

I was writing a script to move all our files in S3 over to Supabase and realized that upload is not working (nor erroring) - while that's maybe not a terribly common use case, doing this server-side would be handy

**Comment by kiwicopple** [comment]
Ah, you know I was supposed to put a big warning in the docs and it completely slipped my mind - last night was a huge rush to get everything out. 

> Let me know if you want a PR, happy to help.

Definitely would love help on this one if you can spare it (everything is in `supabase-js` for now). The key concern was that we didn't want to bloat the library for the browser, so anything we can do to keep the bundle-size down would be great

**Reply by libovness** [reply]
possible to post an example of how to do this server-side (for now) via fetch?

**Reply by libovness** [reply]
I believe it should look something like this?

```
import Image from 'next/image'
import { useState } from 'react'
import useSWR from 'swr'

const fetcher = async (path, token) => {
    const response = await fetch(path, {
        method: 'GET',
        headers: new Headers({ 'Authorization': `Bearer ${token}`}),
    })
    return await response.json();
}

const MemeTemplate = ({template, user, session}) => {

    const [meme, setMeme] = useState()
    const path = `https://[SUPABASE_APP_ID].supabase.co/storage/v1/object/memes/popular/${template.filename}`

    const swr = useSWR(
        session ? [path, session.access_token] : null,
        fetcher
    )
    return meme? (
        <>
            <Image
                src={meme}
                alt=""
                width="290"
                height="290"
            />
            <p>{meme}</p>
        </> 
    ) : ( <div className="avatar no-image" style={{ height: 290, width: 290 }} /> 
    )
}

export default MemeTemplate
```

**Reply by joshnuss** [reply]
@libovness you can upload a file server side using something like this:

```javascript
import { createClient } from '@supabase/supabase-js'
import fs from 'fs'
import FormData from 'form-data'
import fetch from 'node-fetch'

const supabase = createClient('<your-supabase-url>', '<your-supabase-secret-key>')
const bucketName = '<your-bucket-name-here>'
const filePath = 'cat.gif'
const path = `${supabase.storage.url}/object/${bucketName}/${filePath}`
const headers = supabase.storage.headers

const data = await fs.promises.readFile('cat.gif')
const formData = new FormData()

formData.append('', data, filePath)
formData.append('cacheControl', 3600)

const result = await fetch(path, {
  method: 'POST',
  headers,
  body: formData
})

console.log(result)
```

**Reply by IHIutch** [reply]
This pattern doesn't appear to work when hosting on Vercel. I opened an issue here https://github.com/supabase/supabase/issues/1405. It would be nice if the implementation is as easy as you mention!

**Reply by hdoshi2** [reply]
@joshnuss thanks for the post. Is there any way you can breakdown "supabase.storage.headers"? I seem to be getting the following response when I test this route in Postman. Right now I have the following in my header: { 'Authorization': `Bearer ${token}`}
```
{
    "statusCode": "401",
    "error": "Invalid JWT",
    "message": "JWSError JWSInvalidSignature"
}
```

**Comment by inian** [comment, marked as answer]
This feature has been implemented
